src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -206,8 +207,40 @@ public void Dispose()
         [MethodImpl(MethodImplOptions.InternalCall)]
         private static extern IntPtr nInitialize(object? primary, object? secondary);
 
+#if DEBUG
         [MethodImpl(MethodImplOptions.InternalCall)]
         private static extern object? nGetPrimary(IntPtr dependentHandle);
+#else
+        private static unsafe object? nGetPrimary(IntPtr dependentHandle)
+        {
+            // This optimization is the same that is used in GCHandle in RELEASE mode.
+            // This is not used in DEBUG builds as the runtime performs additional checks.
+            // This conversion works because on CoreCLR, DependentHandle is really just a
+            // OBJECTHANDLE value, the same that GCHandle uses. They are defined as follows:
+            //
+            // struct OBJECTHANDLE__
+            // {
+            //     void* unused;
+            // };
+            //
+            // typedef struct OBJECTHANDLE__* OBJECTHANDLE;
+            //
+            // The way this works is that the GC will allocate a special region of memory where
+            // the object reference will be stored (that is, an object* pointer), and the GC will
+            // take care of updating it whenever a compacting collection occurs and the object is
+            // moved. The code below is just doing an indirect memory access to read the pointer
+            // value stored in this memory region, and then converting that to an object reference.
+            // Note that Unsafe.As<IntPtr, object> returns a ref object and does an implicit read.
+            // The code below just compiles down to the following asm on x64:
+            //
+            // mov rax, [rcx]
+            // ret
+            //
+            // That is, it's reading the object* pointer the handle points to, and returning it.
+            // The object* pointer is just expressed object in C#, as it's a reference type.
+            return Unsafe.As<IntPtr, object>(ref *(IntPtr*)dependentHandle);
```

Comment:
nit: This should be `ref *(IntPtr*)(nint)dependentHandle`

Both generate the same code, but casting to `(nint)` first avoids the user-defined call to `void* op_Explicit(IntPtr)` and so results in much smaller IL (and therefore avoids the inliner, additional nodes created by the JIT, etc): https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIBXAOwGcBDAMxjpLoGE6Abyp1RPUkjoBJJhgAKGKACo6AWQAUM+YroATGAAcYTfbIASbE7hgBKOgF4AfHU2yFym/qMnjGC1ZgAbhExENFCCWk3RRUAOVdtKD1DY1M/S11rOycXLXclG3UmbFlPFJ9zDOtgygBfIA

Reply:
Fixed in eb2d2ff18abc64bc81ad141dee3de14a00086c7f 🚀

Reply:
> This might be worth a small analyzer to catch any places we are doing this in the framework
(and similarly for other cases where IntPtr arithmetic is being done but where nint is better)

Or teaching Roslyn to just do this optimization for you to avoid the busy work.

Reply:
-- I'm hoping with the generic math work and `IntPtr` getting all the same operators as `nint` and with the prospect of adding `user-defined checked operators` we can just make `IntPtr`/`nint` have the same relationship as `Int32/int`.
It would require a breaking change (for `IntPtr` not for `nint`) with regards to a couple conversion operators, but I think that will be worth it in the long run.